pkg/store/storepb/custom.go

```diff
@@ -517,3 +517,34 @@ func (c *SeriesStatsCounter) Count(series *Series) {
 		}
 	}
 }
+
+func (m *SeriesRequest) ToPromQL() string {
+	grouping := strings.Join(m.QueryHints.Grouping, ",")
+	matchers := MatchersToString(m.Matchers...)
+	return fmt.Sprintf("%s %s (%s) (%s)", m.QueryHints.Func, m.QueryHints.ByOrWithout(), grouping, matchers)
+}
+
+func (m *QueryHints) IsFuncDistributive() bool {
+	distributiveFuncs := []string{
+		"max",
+		"min",
+		"group",
+		"topk",
+		"bottomk",
```

Comment:
Are we sure, those are distributive?

Reply:
I think they are because the querier will still execute a bottomk on the resultset. In other words
```
bottomk(everything) = bottomk( bottomk(node a), bottomk(node b) )
```
I just don't know if deduplication can affect the result in some way. If the querier sees a union of all nodes, then it will do the right thing.

Reply:
I think that's not true. I wish it was ;p 

All even without `@` modified - there is also this modifier which makes things even more tricky.
Imagine this case:

topk(1, sum(metric1) by (label1)))

```
Node A:

metric1{label1="1", ...} 10
metric1{label1="2", ...} 9

Node B

metric1{label1="1", ...} 1
metric1{label1="2", ...} 9
```

Overall answer is `metric1{label1="2"} 18`, but if we push down we have `metric1{label1="1"} 10` from node A and  `metric1{label1="2"} 9` from node B. when we do aggregated topk 1 it will give `metric1{label1="1"} 10` which is overall incorrect answer.

Reply:
so I think we can pushdown topk and bottomk but only in certain cases

Reply:
As @fpetkovski mentioned offline - it's fine now as we push only first layer of hints

Reply:
@bwplotka Wouldn't the same be true for `max()` as well, since it's mostly equivalent to `topk(1, ...)`? Both would only work if you either don't aggregate away the external labels, or if you first push down the inner `sum()` and feed its merged result to a `topk()` / `max()`. And then the same is true for `bottomk()` / `min()`...